### PR TITLE
refa: reduce usage of send

### PR DESCRIPTION
**Pull request by salzig**
saw issue #1470 and found `lib/portus/migrate.rb` a bit hard to follow. 

Hope you don't mind the refactoring.

### CODE REVIEW

File: lib/portus/migrate.rb [hunk 1]
```diff
@@ -14,29 +14,37 @@ module Migrate
...
-        Rails.logger.warn "The 'x.minutes' format is deprecated for configuration values such " \
-                          "as `jwt_expiration_time`. From now on these values are expected to " \
-                          "be integers representing minutes."
+      raw_value, method = duration.split('.')
+      value = raw_value.to_i
+
+      # When we found a callable method, notify about deprecation
+      unless method.nil?
+        if value.respond_to? method
+          Rails.logger.tagged("deprecated") do
```
Comment (mssola) [inline lib/portus/migrate.rb]: Since you are fixing #1470, raise an exception instead of logging it. I'd create a `DeprecationError` exception class, that can be called like this:

```ruby
raise DeprecationError, "Format 'x.minutes' is no longer supported: time values are expected to be given in minutes."
```

This class would add an "[DEPRECATED]" prefix or something like that to the message.

File: lib/portus/migrate.rb [hunk 2]
```diff
@@ -14,29 +14,37 @@ module Migrate
...
-      Rails.logger.tagged("deprecated") do
-        Rails.logger.warn "The 'x.minutes' format is deprecated for configuration values such " \
-                          "as `jwt_expiration_time`. From now on these values are expected to " \
-                          "be integers representing minutes."
+      raw_value, method = duration.split('.')
+      value = raw_value.to_i
+
+      # When we found a callable method, notify about deprecation
+      unless method.nil?
+        if value.respond_to? method
```
Comment (mssola) [inline lib/portus/migrate.rb]: rubocop will complain about this (it doesn't now because CI is broken, working on that...). Instead you should use a guard clause:

```ruby
return unless value.respond_to? method
```

File: lib/portus/migrate.rb [hunk 3]
```diff
@@ -14,29 +14,37 @@ module Migrate
...
+        end
       end
 
-      # rubocop:disable Lint/Eval
-      eval(duration)
-      # rubocop:enable Lint/Eval
+      # If it's a string containing a positive number, then convert it and return
+      # it. Otherwise, if it has a bad (old) format, then just return the default.
+      if value > 0
+        return value.minutes
```
Comment (mssola) [inline lib/portus/migrate.rb]: Don't use return, simply writing `value.minutes` is enough 😉

File: lib/portus/migrate.rb [hunk 4]
```diff
@@ -14,29 +14,37 @@ module Migrate
...
+      else
+        Rails.logger.warn "Unsupported time format (#{duration}), fallback to default."
+        return default.minutes
```
Comment (mssola) [inline lib/portus/migrate.rb]: Same here

File: lib/portus/migrate.rb (hunk 1 above)
Comment (mssola) [reply lib/portus/migrate.rb]: While you are at it, do the same for the `self.registry_config` method 😁

Comment (salzig) [reply lib/portus/migrate.rb]: wanted to do the refactoring first, as it's kinda hard to read right now. Deprecation Exception would be added in another pull request afterwards.

File: lib/portus/migrate.rb (hunk 3 above)
Comment (salzig) [reply lib/portus/migrate.rb]: you're right.

File: lib/portus/migrate.rb (hunk 4 above)
Comment (salzig) [reply lib/portus/migrate.rb]: you're right.

File: lib/portus/migrate.rb (hunk 1 above)
Comment (salzig) [reply lib/portus/migrate.rb]: btw, is the deprecation only because of the usage of `eval`?

**Comment by mssola** [comment]
> wanted to do the refactoring first, as it's kinda hard to read right now. Deprecation Exception would be added in another pull request afterwards.

I understand that you wanted to do the refactoring first, but I think that you can also fix that issue on this same PR (so also implement the error class). Then this PR (when squashed) would contain two commits:

1. The refactoring itself.
2. The addition of the exception class.

I don't think it would make the PR too complicated to follow honestly.

> btw, is the deprecation only because of the previous usage of eval?

Not really. The deprecation came with the realization that it was just too flexible, and that a mere integer representing time in minutes would suffice. This in turn avoids us of an `eval` call, but it wasn't the primary motivation.

**Comment by mssola** [comment]
@salzig just fixed the CI issues, and upgraded quite some stuff. Due to rubocop's new issues, I had to update the file you were touching and hence the conflict. Sorry about that 😅

**Comment by salzig** [comment]
will rebase later.

File: lib/portus/deprecation_error.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module Portus
+  class DeprecationError < StandardError
```
Comment (mssola) [inline lib/portus/deprecation_error.rb]: Could you add some documentation ? Something simple like: "DeprecationError is the exception to be raised when certain functionality has been deprecated in the latest version and we want to provide migration instructions".

File: lib/portus/migrate.rb [hunk 6]
```diff
@@ -2,6 +2,7 @@ module Portus
   # The Portus::Migrate module implements some methods that are used to handle a
   # change from different versions of Portus where there is an old (and
   # deprecated) way of doing things and a preferred new one.
+
```
Comment (mssola) [inline lib/portus/migrate.rb]: Remove this extra line

File: spec/lib/portus/deprecation_error_spec.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+require "rails_helper"
+
+describe Portus::DeprecationError do
+  let(:message) { "example" }
+  subject(:error) { described_class.new(message) }
+
+  context "#to_s" do
+    subject { error.to_s }
```
Comment (mssola) [inline spec/lib/portus/deprecation_error_spec.rb]: I think it would be simpler to have just one test like this:

```ruby
it "prefixes the message with [DEPRECATED]" do
  is_expected.to eq("[DEPRECATED] #{message}")
end
```

Comment (salzig) [reply spec/lib/portus/deprecation_error_spec.rb]: every project is different :)

**Comment by mssola** [comment]
Also on the second commit, add a "Fixes #1470" on the commit message body 😉

**Comment by salzig** [comment]
the commit messages contain the progress (commits before squash). Do you want them to be removed?

**Comment by mssola** [comment]
Oh I see. Yes please, remove those messages. Just leave a single and cohesive message explaining what's in the commit.

**Comment by salzig** [comment]
any further change required?
